logtracing/span.go

```diff
@@ -0,0 +1,114 @@
+package logtracing
+
+import (
+	"bytes"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"sync"
+	"time"
+)
+
+type span struct {
+	parent *span
+
+	traceID     TraceID
+	spanID      SpanID
+	spanContext string
+
+	startTime time.Time
+	endTime   time.Time
+
+	err   error
+	panic interface{}
+
+	keyvals []interface{}
+	mu      sync.Mutex
+}
+
+func (s *span) IsRecording() bool {
+	return s.endTime.IsZero()
+}
+
+func (s *span) RecordError(err error) {
+	s.err = err
+}
+
+func (s *span) RecordPanic(panic interface{}) {
+	s.panic = panic
+}
+
+func (s *span) End() {
+	if !s.IsRecording() {
+		return
+	}
+
+	s.endTime = time.Now()
+}
+
+func (s *span) Duration() time.Duration {
+	if s.IsRecording() {
+		return 0
+	}
+	return s.endTime.Sub(s.startTime)
+}
+
+var ErrMissingValue = errors.New("(MISSING)")
+
+func (s *span) AppendKVs(keyvals ...interface{}) {
+	if len(keyvals)%2 != 0 {
+		keyvals = append(keyvals, ErrMissingValue)
+	}
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	s.keyvals = append(s.keyvals, keyvals...)
+}
+
+// TraceID is a unique identity of a trace.
+// nolint:revive // revive complains about stutter of `trace.TraceID`.
+type TraceID [16]byte
```

Comment:
**About trace ID and span ID**
(I did some benchmarks -> found `logtracing` is slower than `opencensus` -> referenced the source code of `opencensus`)

Two main reasons:

1. It's the recommendation in [W3C Trace Context](https://www.w3.org/TR/trace-context/#trace-id). I haven't dug too deep, I thought it might be good for integrating with other tracing frameworks, such as OpenTelemetry and OpenCensus
2. The `randomIDGenerator` is faster than `uuid.New`. Well, 31ns/op vs 700ns/op...

If `uuid` is not MUST have, I'd like to use these new byte-array IDs.

Reply:
👍